### PR TITLE
Fix Lifecycle discovery of @OnStage in Provider-provisioned Guice bindings

### PR DESCRIPTION
**Pull request by stevenschlansker**

### CODE REVIEW

File: guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java [hunk 1]
```diff
@@ -68,19 +66,14 @@
 
                 LOG.trace("Will invoke {} on {}", m, onStage.value());
 
-                encounter.register(new InjectionListener<I>() {
-                    @Override
-                    public void afterInjection(I injectee) {
```
Comment (pennello) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: So is it correct to say that the sequencing was just wrong here before?  That is, we shouldn't have been performing this invocation logic after injection?

File: guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java [hunk 2]
```diff
@@ -41,7 +38,7 @@
  * {@link Lifecycle} so that they are run.
  */
 @NotThreadSafe
-class LifecycleAnnotationFinder implements TypeListener {
+class LifecycleAnnotationFinder implements ProvisionListener {
```
Comment (pennello) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: The provision listener behavior is different from the type listener.  Is there any functionality that we will now be missing out on by using this interface instead?

Comment (stevenschlansker) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: Not that I'm aware of.  but I'm no expert on Guice internals... :/
We have good test coverage and they continue to pass, at least?

File: guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java (hunk 1 above)
Comment (stevenschlansker) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: No, it has to be after injection.  And in fact it still is -- you'll notice that we call `provision.provision()` up above.  So this is a different interface to access a similar point in the injected object lifecycle.

Comment (pennello) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: Yes, I did note the call above.  So what logically changed here then?  If we're doing the same thing, in the same order, but just via a slightly different method, how does this current implementation fix things?

Comment (stevenschlansker) [inline guice/src/main/java/com/opentable/lifecycle/guice/LifecycleAnnotationFinder.java]: the previous `TypeListener` only got the type of the binding as declared -- not the actual type used.

So in the case of

```java
class Foo {}
class Bar extends Foo { @OnStage someMethod() { } }
bind(Foo.class).toProvider(MakeABar.class);
```

the `TypeListener` would scan `Foo`, see no `@OnStage` methods, and never create an `InjectionListener`.  The new `ProvisionListener` (which did not exist at the time this code was authored) does not have this problem, as it is given the actual provisioned instance, not just the statically declared type.

**Comment by benhardy** [comment]
Looks good to me. Well spotted.

**Pull request merged**